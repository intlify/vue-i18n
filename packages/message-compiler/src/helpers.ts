export const enum HelperNameMap {
  LIST = 'list',
  NAMED = 'named',
  PLURAL = 'plural',
  LINKED = 'linked',
  MESSAGE = 'message',
  TYPE = 'type',
  INTERPOLATE = 'interpolate',
  NORMALIZE = 'normalize'
}
