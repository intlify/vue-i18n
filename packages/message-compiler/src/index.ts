export * from './options'
export * from './errors'
export * from './helpers'
export * from './location'
export * from './compiler'
