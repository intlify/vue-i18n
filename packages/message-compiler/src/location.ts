// The source range.
// The `start` is inclusive and `end` is exclusive.
// [start, end)
export interface SourceLocation {
  start: Position
  end: Position
  source?: string
}

export const LocationStub: SourceLocation = {
  start: { line: 1, column: 1, offset: 0 },
  end: { line: 1, column: 1, offset: 0 }
}

export interface Position {
  offset: number // from start of content (file)
  line: number
  column: number
}

export function createPosition(
  line: number,
  column: number,
  offset: number
): Position {
  return { line, column, offset } as Position
}

export function createLocation(
  start: Position,
  end: Position,
  source?: string
): SourceLocation {
  const loc = { start, end } as SourceLocation
  if (source != null) {
    loc.source = source
  }
  return loc
}
