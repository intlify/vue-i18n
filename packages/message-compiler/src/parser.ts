import { Position, createLocation, SourceLocation } from './location'
import { ParserOptions } from './options'
import { createCompileError, CompileErrorCodes } from './errors'
import { Tokenizer, createTokenizer, TokenTypes, Token } from './tokenizer'
import { assign } from '@intlify/shared'

export const enum NodeTypes {
  Resource, // 0
  Plural,
  Message,
  Text,
  Named,
  List, // 5
  Linked,
  LinkedKey,
  LinkedModifier,
  Literal
}

// not containing whitespace or control characters
export type Identifier = string

export interface Node {
  type: NodeTypes
  start: number
  end: number
  loc?: SourceLocation
}

export interface ResourceNode extends Node {
  type: NodeTypes.Resource
  body: MessageNode | PluralNode
  helpers?: string[]
}

export interface PluralNode extends Node {
  type: NodeTypes.Plural
  cases: MessageNode[]
}

export interface MessageNode extends Node {
  type: NodeTypes.Message
  items: MessageElementNode[]
}

type MessageElementNode =
  | TextNode
  | NamedNode
  | ListNode
  | LiteralNode
  | LinkedNode

export interface TextNode extends Node {
  type: NodeTypes.Text
  value: string
}

export interface NamedNode extends Node {
  type: NodeTypes.Named
  key: Identifier
}

export interface ListNode extends Node {
  type: NodeTypes.List
  index: number
}

export interface LiteralNode extends Node {
  type: NodeTypes.Literal
  value: string
}

export interface LinkedNode extends Node {
  type: NodeTypes.Linked
  modifier?: LinkedModifierNode
  key: LinkedKeyNode | NamedNode | ListNode | LiteralNode
}

export interface LinkedKeyNode extends Node {
  type: NodeTypes.LinkedKey
  value: string
}

export interface LinkedModifierNode extends Node {
  type: NodeTypes.LinkedModifier
  value: Identifier
}

export interface Parser {
  parse(source: string): ResourceNode
}

export const ERROR_DOMAIN = 'parser'

// Backslash backslash, backslash quote, uHHHH, UHHHHHH.
const KNOWN_ESCAPES = /(?:\\\\|\\'|\\u([0-9a-fA-F]{4})|\\U([0-9a-fA-F]{6}))/g

function fromEscapeSequence(
  match: string,
  codePoint4: string,
  codePoint6: string
): string {
  switch (match) {
    case `\\\\`:
      return `\\`
    case `\\\'`:
      return `\'`
    default: {
      const codePoint = parseInt(codePoint4 || codePoint6, 16)
      if (codePoint <= 0xd7ff || codePoint >= 0xe000) {
        return String.fromCodePoint(codePoint)
      }
      // invalid ...
      // Replace them with U+FFFD REPLACEMENT CHARACTER.
      return 'ï¿½'
    }
  }
}

export function createParser(options: ParserOptions = {}): Parser {
  const location = options.location !== false

  const { onError } = options
  function emitError(
    tokenzer: Tokenizer,
    code: CompileErrorCodes,
    start: Position,
    offset: number,
    ...args: unknown[]
  ): void {
    const end = tokenzer.currentPosition()
    end.offset += offset
    end.column += offset
    if (onError) {
      const loc = createLocation(start, end)
      const err = createCompileError(code, loc, {
        domain: ERROR_DOMAIN,
        args
      })
      onError(err)
    }
  }

  function startNode(type: NodeTypes, offset: number, loc: Position): Node {
    const node = {
      type,
      start: offset,
      end: offset
    } as Node

    if (location) {
      node.loc = { start: loc, end: loc }
    }

    return node
  }

  function endNode(
    node: Node,
    offset: number,
    pos: Position,
    type?: NodeTypes
  ): void {
    node.end = offset

    if (type) {
      node.type = type
    }

    if (location && node.loc) {
      node.loc.end = pos
    }
  }

  function parseText(tokenizer: Tokenizer, value: string): TextNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.Text,
      context.offset,
      context.startLoc
    ) as TextNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseList(tokenizer: Tokenizer, index: string): ListNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.List, offset, loc) as ListNode
    node.index = parseInt(index, 10)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseNamed(tokenizer: Tokenizer, key: string): NamedNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Named, offset, loc) as NamedNode
    node.key = key
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLiteral(tokenizer: Tokenizer, value: string): LiteralNode {
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get brace left loc
    const node = startNode(NodeTypes.Literal, offset, loc) as LiteralNode
    node.value = value.replace(KNOWN_ESCAPES, fromEscapeSequence)
    tokenizer.nextToken() // skip brach right
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinkedModifier(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedModifierNode
  } {
    const token = tokenizer.nextToken()
    const context = tokenizer.context()
    const { lastOffset: offset, lastStartLoc: loc } = context // get linked dot loc
    const node = startNode(
      NodeTypes.LinkedModifier,
      offset,
      loc
    ) as LinkedModifierNode
    if (token.type !== TokenTypes.LinkedModifier) {
      // empty modifier
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_MODIFIER,
        context.lastStartLoc,
        0
      )
      node.value = ''
      endNode(node, offset, loc)
      return {
        nextConsumeToken: token,
        node
      }
    }
    // check token
    if (token.value == null) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    node.value = token.value || ''
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node
    }
  }

  function parseLinkedKey(tokenizer: Tokenizer, value: string): LinkedKeyNode {
    const context = tokenizer.context()
    const node = startNode(
      NodeTypes.LinkedKey,
      context.offset,
      context.startLoc
    ) as LinkedKeyNode
    node.value = value
    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseLinked(
    tokenizer: Tokenizer
  ): {
    nextConsumeToken?: Token
    node: LinkedNode
  } {
    const context = tokenizer.context()
    const linkedNode = startNode(
      NodeTypes.Linked,
      context.offset,
      context.startLoc
    ) as LinkedNode

    let token = tokenizer.nextToken()
    if (token.type === TokenTypes.LinkedDot) {
      const parsed = parseLinkedModifier(tokenizer)
      linkedNode.modifier = parsed.node
      token = parsed.nextConsumeToken || tokenizer.nextToken()
    }

    // asset check token
    if (token.type !== TokenTypes.LinkedDelimiter) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        getTokenCaption(token)
      )
    }
    token = tokenizer.nextToken()

    // skip brace left
    if (token.type === TokenTypes.BraceLeft) {
      token = tokenizer.nextToken()
    }

    switch (token.type) {
      case TokenTypes.LinkedKey:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLinkedKey(tokenizer, token.value || '')
        break
      case TokenTypes.Named:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseNamed(tokenizer, token.value || '')
        break
      case TokenTypes.List:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseList(tokenizer, token.value || '')
        break
      case TokenTypes.Literal:
        if (token.value == null) {
          emitError(
            tokenizer,
            CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
            context.lastStartLoc,
            0,
            getTokenCaption(token)
          )
        }
        linkedNode.key = parseLiteral(tokenizer, token.value || '')
        break
      default:
        // empty key
        emitError(
          tokenizer,
          CompileErrorCodes.UNEXPECTED_EMPTY_LINKED_KEY,
          context.lastStartLoc,
          0
        )
        const nextContext = tokenizer.context()
        const emptyLinkedKeyNode = startNode(
          NodeTypes.LinkedKey,
          nextContext.offset,
          nextContext.startLoc
        ) as LinkedKeyNode
        emptyLinkedKeyNode.value = ''
        endNode(emptyLinkedKeyNode, nextContext.offset, nextContext.startLoc)
        linkedNode.key = emptyLinkedKeyNode
        endNode(linkedNode, nextContext.offset, nextContext.startLoc)
        return {
          nextConsumeToken: token,
          node: linkedNode
        }
        break
    }

    endNode(linkedNode, tokenizer.currentOffset(), tokenizer.currentPosition())
    return {
      node: linkedNode
    }
  }

  function parseMessage(tokenizer: Tokenizer): MessageNode {
    const context = tokenizer.context()
    const startOffset =
      context.currentType === TokenTypes.Pipe
        ? tokenizer.currentOffset()
        : context.offset
    const startLoc =
      context.currentType === TokenTypes.Pipe
        ? context.endLoc
        : context.startLoc
    const node = startNode(
      NodeTypes.Message,
      startOffset,
      startLoc
    ) as MessageNode
    node.items = []

    let nextToken: Token | null = null
    do {
      const token = nextToken || tokenizer.nextToken()
      nextToken = null
      switch (token.type) {
        case TokenTypes.Text:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseText(tokenizer, token.value || ''))
          break
        case TokenTypes.List:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseList(tokenizer, token.value || ''))
          break
        case TokenTypes.Named:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseNamed(tokenizer, token.value || ''))
          break
        case TokenTypes.Literal:
          if (token.value == null) {
            emitError(
              tokenizer,
              CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
              context.lastStartLoc,
              0,
              getTokenCaption(token)
            )
          }
          node.items.push(parseLiteral(tokenizer, token.value || ''))
          break
        case TokenTypes.LinkedAlias:
          const parsed = parseLinked(tokenizer)
          node.items.push(parsed.node)
          nextToken = parsed.nextConsumeToken || null
          break
      }
    } while (
      context.currentType !== TokenTypes.EOF &&
      context.currentType !== TokenTypes.Pipe
    )

    // adjust message node loc
    const endOffset =
      context.currentType === TokenTypes.Pipe
        ? context.lastOffset
        : tokenizer.currentOffset()
    const endLoc =
      context.currentType === TokenTypes.Pipe
        ? context.lastEndLoc
        : tokenizer.currentPosition()

    endNode(node, endOffset, endLoc)
    return node
  }

  function parsePlural(
    tokenizer: Tokenizer,
    offset: number,
    loc: Position,
    msgNode: MessageNode
  ): PluralNode {
    const context = tokenizer.context()
    let hasEmptyMessage = msgNode.items.length === 0

    const node = startNode(NodeTypes.Plural, offset, loc) as PluralNode
    node.cases = []
    node.cases.push(msgNode)

    do {
      const msg = parseMessage(tokenizer)
      if (!hasEmptyMessage) {
        hasEmptyMessage = msg.items.length === 0
      }
      node.cases.push(msg)
    } while (context.currentType !== TokenTypes.EOF)

    if (hasEmptyMessage) {
      emitError(
        tokenizer,
        CompileErrorCodes.MUST_HAVE_MESSAGES_IN_PLURAL,
        loc,
        0
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  function parseResource(tokenizer: Tokenizer): MessageNode | PluralNode {
    const context = tokenizer.context()
    const { offset, startLoc } = context

    const msgNode = parseMessage(tokenizer)
    if (context.currentType === TokenTypes.EOF) {
      return msgNode
    } else {
      return parsePlural(tokenizer, offset, startLoc, msgNode)
    }
  }

  function parse(source: string): ResourceNode {
    const tokenizer = createTokenizer(source, assign({}, options))
    const context = tokenizer.context()

    const node = startNode(
      NodeTypes.Resource,
      context.offset,
      context.startLoc
    ) as ResourceNode
    if (location && node.loc) {
      node.loc.source = source
    }
    node.body = parseResource(tokenizer)

    // assert whether achieved to EOF
    if (context.currentType !== TokenTypes.EOF) {
      emitError(
        tokenizer,
        CompileErrorCodes.UNEXPECTED_LEXICAL_ANALYSIS,
        context.lastStartLoc,
        0,
        source[context.offset] || ''
      )
    }

    endNode(node, tokenizer.currentOffset(), tokenizer.currentPosition())
    return node
  }

  return { parse }
}

function getTokenCaption(token: Token) {
  if (token.type === TokenTypes.EOF) {
    return 'EOF'
  }
  const name = (token.value || '').replace(/\r?\n/gu, '\\n')
  return name.length > 10 ? name.slice(0, 9) + 'â¦' : name
}
