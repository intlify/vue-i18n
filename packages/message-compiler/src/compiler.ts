import { CompileOptions } from './options'
import { createParser } from './parser'
import { transform } from './transformer'
import { generate, CodeGenResult } from './generator'
import { assign } from '@intlify/shared'

export function baseCompile(
  source: string,
  options: CompileOptions = {}
): CodeGenResult {
  const assignedOptions = assign({}, options)

  // parse source codes
  const parser = createParser(assignedOptions)
  const ast = parser.parse(source)

  // transform ASTs
  transform(ast, assignedOptions)

  // generate javascript codes
  return generate(ast, assignedOptions)
}
