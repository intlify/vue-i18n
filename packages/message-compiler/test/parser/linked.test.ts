import { createParser, NodeTypes, MessageNode } from '../../src/parser'

let spy: any // eslint-disable-line @typescript-eslint/no-explicit-any
beforeEach(() => {
  spy = jest.fn()
})

describe('basic', () => {
  test(`hi @:name !`, () => {
    const text = `hi @:name !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.LinkedKey,
          value: 'name'
        }
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('literal key', () => {
  test(`hi @:{'hello world'} !`, () => {
    const text = `hi @:{'hello world'} !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.Literal,
          value: 'hello world'
        }
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('terminated with new line', () => {
  test(`hi @:{name}\n !`, () => {
    const text = `hi @:{name}\n !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.Named,
          key: 'name'
        }
      },
      {
        type: NodeTypes.Text,
        value: '\n !'
      }
    ])
  })
})

describe('modifier', () => {
  test(`hi @.upper:name !`, () => {
    const text = `hi @.upper:{name} !`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(3)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.Named,
          key: 'name'
        },
        modifier: {
          type: NodeTypes.LinkedModifier,
          value: 'upper'
        }
      },
      {
        type: NodeTypes.Text,
        value: ' !'
      }
    ])
  })
})

describe('multiple', () => {
  test(`hi @:{name} @:{0}!`, () => {
    const text = `hi @:{name} @:{0}!`
    const parser = createParser({ onError: spy })
    const ast = parser.parse(text)

    expect(ast).toMatchSnapshot()
    expect(spy).not.toHaveBeenCalled()

    expect(ast.type).toEqual(NodeTypes.Resource)
    expect(ast.body.type).toEqual(NodeTypes.Message)
    const message = ast.body as MessageNode
    expect(message.items).toHaveLength(5)
    expect(message.items).toMatchObject([
      {
        type: NodeTypes.Text,
        value: 'hi '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.Named,
          key: 'name'
        }
      },
      {
        type: NodeTypes.Text,
        value: ' '
      },
      {
        type: NodeTypes.Linked,
        key: {
          type: NodeTypes.List,
          index: 0
        }
      },
      {
        type: NodeTypes.Text,
        value: '!'
      }
    ])
  })
})

describe('edge cases', () => {
  test.todo(`hi @:{ name } !`)
  test.todo(`hi @:{ 'name' } !`)
  test.todo(`hi @:{ {name} } !`)
  test.todo(`foo@bar.com`)
  test.todo(`hi @:\nname !`)
  test.todo(`hi @ :name !`)
  test.todo(`hi @. {name} !`)
  test.todo(`hi @.\n{name} !`)
  test.todo(`hi @.upper\n{name} !`)
  test.todo(`hi @.upper {name} !`)
  test.todo(`hi @:\n{name} !`)
  test.todo(`hi @: {'name'} !`)
  test.todo(`hi @. {name} !`)
  test.todo(`hi @\. upper\n: {'name'}\n !`)
})
