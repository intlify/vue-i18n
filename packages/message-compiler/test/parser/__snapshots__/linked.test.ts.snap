// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic hi @:name ! 1`] = `
Object {
  "body": Object {
    "end": 11,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 9,
        "key": Object {
          "end": 9,
          "loc": Object {
            "end": Object {
              "column": 10,
              "line": 1,
              "offset": 9,
            },
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "start": 5,
          "type": 7,
          "value": "name",
        },
        "loc": Object {
          "end": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 11,
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 10,
            "line": 1,
            "offset": 9,
          },
        },
        "start": 9,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 12,
        "line": 1,
        "offset": 11,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 11,
  "loc": Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "source": "hi @:name !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`literal key hi @:{'hello world'} ! 1`] = `
Object {
  "body": Object {
    "end": 22,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 20,
        "key": Object {
          "end": 20,
          "loc": Object {
            "end": Object {
              "column": 21,
              "line": 1,
              "offset": 20,
            },
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "start": 5,
          "type": 9,
          "value": "hello world",
        },
        "loc": Object {
          "end": Object {
            "column": 21,
            "line": 1,
            "offset": 20,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 22,
        "loc": Object {
          "end": Object {
            "column": 23,
            "line": 1,
            "offset": 22,
          },
          "start": Object {
            "column": 21,
            "line": 1,
            "offset": 20,
          },
        },
        "start": 20,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 23,
        "line": 1,
        "offset": 22,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 22,
  "loc": Object {
    "end": Object {
      "column": 23,
      "line": 1,
      "offset": 22,
    },
    "source": "hi @:{'hello world'} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`modifier hi @.upper:name ! 1`] = `
Object {
  "body": Object {
    "end": 19,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 17,
        "key": Object {
          "end": 17,
          "key": "name",
          "loc": Object {
            "end": Object {
              "column": 18,
              "line": 1,
              "offset": 17,
            },
            "start": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
          },
          "start": 11,
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "modifier": Object {
          "end": 10,
          "loc": Object {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "start": Object {
              "column": 5,
              "line": 1,
              "offset": 4,
            },
          },
          "start": 4,
          "type": 8,
          "value": "upper",
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 19,
        "loc": Object {
          "end": Object {
            "column": 20,
            "line": 1,
            "offset": 19,
          },
          "start": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
        },
        "start": 17,
        "type": 3,
        "value": " !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 20,
        "line": 1,
        "offset": 19,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 19,
  "loc": Object {
    "end": Object {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "source": "hi @.upper:{name} !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`multiple hi @:{name} @:{0}! 1`] = `
Object {
  "body": Object {
    "end": 18,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 11,
        "key": Object {
          "end": 11,
          "key": "name",
          "loc": Object {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "start": 5,
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 12,
        "loc": Object {
          "end": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
          "start": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
        },
        "start": 11,
        "type": 3,
        "value": " ",
      },
      Object {
        "end": 17,
        "key": Object {
          "end": 17,
          "index": 0,
          "loc": Object {
            "end": Object {
              "column": 18,
              "line": 1,
              "offset": 17,
            },
            "start": Object {
              "column": 15,
              "line": 1,
              "offset": 14,
            },
          },
          "start": 14,
          "type": 5,
        },
        "loc": Object {
          "end": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
          "start": Object {
            "column": 13,
            "line": 1,
            "offset": 12,
          },
        },
        "start": 12,
        "type": 6,
      },
      Object {
        "end": 18,
        "loc": Object {
          "end": Object {
            "column": 19,
            "line": 1,
            "offset": 18,
          },
          "start": Object {
            "column": 18,
            "line": 1,
            "offset": 17,
          },
        },
        "start": 17,
        "type": 3,
        "value": "!",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 19,
        "line": 1,
        "offset": 18,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 18,
  "loc": Object {
    "end": Object {
      "column": 19,
      "line": 1,
      "offset": 18,
    },
    "source": "hi @:{name} @:{0}!",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;

exports[`terminated with new line hi @:{name}
 ! 1`] = `
Object {
  "body": Object {
    "end": 14,
    "items": Array [
      Object {
        "end": 3,
        "loc": Object {
          "end": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
          "start": Object {
            "column": 1,
            "line": 1,
            "offset": 0,
          },
        },
        "start": 0,
        "type": 3,
        "value": "hi ",
      },
      Object {
        "end": 11,
        "key": Object {
          "end": 11,
          "key": "name",
          "loc": Object {
            "end": Object {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": Object {
              "column": 6,
              "line": 1,
              "offset": 5,
            },
          },
          "start": 5,
          "type": 4,
        },
        "loc": Object {
          "end": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
          "start": Object {
            "column": 4,
            "line": 1,
            "offset": 3,
          },
        },
        "start": 3,
        "type": 6,
      },
      Object {
        "end": 14,
        "loc": Object {
          "end": Object {
            "column": 3,
            "line": 2,
            "offset": 14,
          },
          "start": Object {
            "column": 12,
            "line": 1,
            "offset": 11,
          },
        },
        "start": 11,
        "type": 3,
        "value": "
 !",
      },
    ],
    "loc": Object {
      "end": Object {
        "column": 3,
        "line": 2,
        "offset": 14,
      },
      "start": Object {
        "column": 1,
        "line": 1,
        "offset": 0,
      },
    },
    "start": 0,
    "type": 2,
  },
  "end": 14,
  "loc": Object {
    "end": Object {
      "column": 3,
      "line": 2,
      "offset": 14,
    },
    "source": "hi @:{name}
 !",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "start": 0,
  "type": 0,
}
`;
